add a chapter to help the UI Port (#525)
MicroEJ/docs: VEEPortingGuide/uiPortSimulation.rst

.. _ui_port_sim:

==========
Simulation
==========

Principle
=========

> The simulation involves creating or extending a Front Panel project to be UI Pack compatible.

Reviewer: ```suggestion
The simulation part of the UI port requires the creation (or extension) of a Front Panel project which is compatible with the UI Pack.
```

First, if no Front Panel project exists, follow the steps described here: :ref:`section_frontpanel`.
Then, follow the next chapters to extend the Front Panel project with UI Pack notions.

Project Extension
=================

The Front Panel project must depend on the UI Pack.
Add the following dependency to the Front Panel ivy file:

.. code-block:: xml

   <dependency org="com.microej.pack.ui" name="ui-pack" rev="[UI Pack version]">
      <artifact name="frontpanel" type="jar"/>
   </dependency>

See :ref:`section_ui_simulation` for more information about the Front Panel project dependencies.

LEDs
====

When the VEE Port Configuration project LEDs module is checked, the Front Panel project should add a widget ``LED`` for each led.

1. With an image editor, create an image for the LED off and an image for the LED on. Both images must have the same size. 
2. Create a couple of images for each LED.
3. In the Front Panel description file, add this line for each LED:

.. code-block:: xml 

   <ej.fp.widget.LED label="0" x="170" y="753" ledOff="Led-0.png" ledOn="Led-GREEN.png" overlay="false"/>

The label must have an integer value from 0 to ``NUMBER_OF_LEDS - 1``.

> The ``ej.microui.led.Leds`` class used this value as the led identifier in ``setLedOff(int ledId)``, ``setLedOn(int ledId)``, and other methods of the class.

Reviewer: ```suggestion
The ``ej.microui.led.Leds`` class uses this value as the LED identifier in ``setLedOff(int ledId)``, ``setLedOn(int ledId)``, and other methods of the class.
```

Buttons
=======

> The widget ``Button`` should simulate each hardware button (if any).

Reviewer: ```suggestion
The widget ``Button`` can simulate any hardware button.
```

.. code-block:: java

   Buttons[] buttonsHandlers = (Buttons[]) EventGenerator.get(Buttons.class);
   for (EventGenerator buttonsHandler : generators) {
      buttonsHandler.setEventHandler(this);
   }

Here is an example of a handler:

.. code-block:: java

   @Override
   public boolean handleEvent(int event) {

      // get the event's data
      int data = Event.getData(event);

      String state = null;

      // print its state(s)
      if (Buttons.isPressed(data)) {
         state = "pressed ";
      }
      if (Buttons.isReleased(data)) {
         state = "released ";
      }
      if (Buttons.isRepeated(data)) {
         state = "repeated ";
      }
      if (Buttons.isLong(data)) {
         state = "long ";
      }
      if (Buttons.isClicked(data)) {
         state = "clicked ";
      }
      if (Buttons.isDoubleClicked(data)) {
         state = "double-clicked ";
      }

      if (state != null) {
         System.out.print("button\t\t");

         // get the button's id
         int id = Buttons.getButtonId(data);
         System.out.print(id+" ");
         System.out.println(state);
      }

      return true;
   }

Button to Command Event
-----------------------

> To be more independent, the widget should send a MicroUI Command event, which is more generic, and the application code is better.

Reviewer: ```suggestion
A recommended approach is to favor Command events over Buttons events. MicroUI Command events are more generic because not they are not tied to a hardware component like a physical button. Command events make the application code more flexible to hardware changes.
```

> For instance, instead of reacting on Button event 0, the application will respond on Command ``Enter`` or ``Up``. 

Reviewer: ```suggestion
For instance, instead of reacting to Button event 0, the application will respond to Command event ``Enter`` or ``Up``. 
```

> The application does not care about the source of the Command event: maybe it is the button 0, 1, 10, or any other input device.

Reviewer: ```suggestion
The application does not care about the source of the Command event: it may be the button 0, 1, 10, or any other input device.
```

      public int getCommand(int buttonId) {
         // same command as EmbJPF (see buttons_listener.c)
         switch (buttonId) {
         default:
         case 0:
            return EventCommand.ESC;
         case 1:
            return EventCommand.MENU;
         }
      }

      @Override
      public void press(Button widget) {
         EventCommand.sendEvent(getCommand(widget.getID()));
      }

      @Override
      public void release(Button widget) {
         // nothing to send
      }
   }

The application code becomes:

.. code-block:: java

   // [...]

      Command[] commandHandlers = (Command[]) EventGenerator.get(Command.class);
      for (EventGenerator commandHandler : generators) {
         commandHandler.setEventHandler(this);
      }

   // [...]

   @Override
   public boolean handleEvent(int event) {

      // get the event's data
      int data = Event.getData(event);

      switch (data) {
      case Command.ESC:
         System.out.println("ESC");
         break;
      case Command.BACK:
         System.out.println("BACK");
         break;

      // [...]
   }

Touch Panel
===========

> Contrary to the other input devices, no image is required: a touch panel is over the display.

Reviewer: ```suggestion
Contrary to the other input devices, no image is required because a touch panel covers the display area.
```

1. Retrieve the display size in pixels.
2. In the Front Panel description file, add this line:

.. code-block:: xml 

   <ej.fp.widget.Pointer x="185" y="395" width="480" height="272" touch="true"/>

By default, the widget sends a MicroUI Pointer event to the Pointer Event Generator, whose name is ``TOUCH`` (a touch panel is considered a Pointer with only dragged events).
To target another Pointer Event Generator, refer to the chapter :ref:`section_ui_simulation_input`.

> A snippet of application code:

Reviewer: ```suggestion
A snippet of application code that handles Pointer events:
```

.. code-block:: java

   // [...]

      Pointer[] pointerHandlers = (Pointer[]) EventGenerator.get(Pointer.class);
      for (EventGenerator pointerHandler : generators) {
         pointerHandler.setEventHandler(this);
      }

   // [...]

   @Override
   public boolean handleEvent(int event) {
      Pointer pointer = (Pointer) Event.getGenerator(event);
      int x = pointer.getX();
      int y = pointer.getY();
      System.out.println("(" + x + "," + y + ")");
   }

Display
=======

The widget Display features a lot of options to simulate the hardware specificities.

1. Retrieve the display size in pixels.
2. In the Front Panel description file, add this line:

.. code-block:: xml 

   <ej.fp.widget.Display x="185" y="395" width="480" height="272"/>

For more information, refer to the java-doc of the widget Display and the chapter :ref:`section_ui_simulation_display`.

Build
=====

> Once the Front Panel project is created or modified, the VEE Port must be built again as the front panel is built simultaneously with the VEE Port; see :ref:`platform_build`.

Reviewer: ```suggestion
Once the Front Panel project is created or modified, the VEE Port must be built again (the front panel is built simultaneously with the VEE Port; see :ref:`platform_build`).
```
